0010-zkvm-keccak.md

```diff
@@ -0,0 +1,788 @@
+| `Reset64` | sparse0  | sparse1  | sparse2  | sparse3  | reset1_0 | reset1_1 | word | reset2_0 | reset2_1 | reset3_0 | reset3_1 | dense0 | dense1 | 
+| `Zero`    | reset0_0 | reset0_1 | reset0_2 | reset0_3 | reset1_2 | reset1_3 |      | reset2_2 | reset2_3 | reset3_2 | reset3_3 | dense2 | dense3 |
+
+| `Not64` | `0*` | `1*` | `2*` | `3*` | `4*` | `5*` | `6*` | `7*` |
+| ------- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- |
+| `Curr`  | x0   | x1   | x2   | x3   | not0 | not1 | not2 | not3 | 
+
+| `Rot64` | `0*`   | `1*`   | `2!` | `3!` | `4!` | `5!` | `6!` | `7!` | `8!` | `9!` | `10!` | `11!` | `12!` | `13!` | 
+| ------- | ------ | ------ | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ----- | ----- | ----- | ----- |
+| `Curr`  | input  | output | quo0 | quo1 | quo2 | quo3 | rem0 | rem1 | rem2 | rem3 | bnd0  | bnd1  | bnd2  | bnd3  |
+
+with one coefficient set to `2^{offset}`.
+
+## Prior art
+
+The current Keccak PoC in SnarkyML was introduced to support Ethereum primitives in MINA zkApps. Due to this blockchain's design, the gadget needed to be compatible with Kimchi: a Plonk-like SNARK instantiated with Pasta curves for Pickles recursion and IPA commitents. This means that the design choices for that gadget were determined by some features of this proof system, such as: 15-column width witness, 7 permutable witness cells per row, access to the current and next rows, up to 4 lookups per row, access to 4-bit XOR lookup table, and less than $2^{16}$ rows. As a result, proving the Keccak hash of a message of 1 block length (up to 1080 bits) took ~15k rows.
+
+| Version  | Columns | Rows / block | Lookups / block | 
+|----------|---------|--------------|-----------------|
+| Old PoC  | 15      | $24\times(125+100+40+125+75+287.5+5)=18,180$ |$24\times(400+320+140+400+300+800+16)=57,024$ |   
+
+## Unresolved questions
+
+* During the review of this RFC:
+    * Resolve how to deal with input and output.
+
+* During the implementation of this RFC:
+    * obtain exact measurements of the number of rows, columns, constraints, lookups, seconds, required per block hash;
+    * if the endianness of the target input format is little endian, it will need to be transformed into big endian;
+    * pack the bytestring given in chunks of 4 bytes.
+
+* Future work: 
+    * support SHA3 (NIST variant of Keccak), different output lengths, and different state widths;
+    * improve the lookup argument to reuse the `Reset` table also for single-column lookups. Perhaps this can work adding some logic such as performing lookups modulo a constant ($2^{64}$ in this case), or shifting right by a given number of positions ($64$ in this case);
+    * optimize the `KeccakRound` to remove redundant intermediate states (some preliminar ideas have been performed);
+    * optimize range checks or think of a different approach towards rotation.
+
+
+## Appendix
```

Comment:
I think adding a link to the keccak.team website might be good. They have good resources and the pseudo code on the specs summary is clear.